src/functions/asyncCredential/asyncCredentialHandler.test.ts

```diff
@@ -1,14 +1,640 @@
+
+          dependencies.tokenService = () => new MockTokenSeviceValidSignature();
+
+          const result: APIGatewayProxyResult = await lambdaHandler(
+            event,
+            dependencies,
+          );
+
+          expect(result).toStrictEqual({
+            headers: { "Content-Type": "application/json" },
+            statusCode: 400,
+            body: JSON.stringify({
+              error: "invalid_token",
+              error_description: "Missing client_id claim",
+            }),
+          });
+        });
+      });
+    });
+
+    describe("aud claim validation", () => {
+      describe("Given aud (audience) is missing", () => {
+        it("Returns a log", async () => {
+          const event = buildRequest({
+            headers: { Authorization: `Bearer ${mockJwtNoAud}` },
+          });
+
+          dependencies.tokenService = () => new MockTokenSeviceValidSignature();
+
+          const result: APIGatewayProxyResult = await lambdaHandler(
+            event,
+            dependencies,
+          );
+
+          expect(result).toStrictEqual({
+            headers: { "Content-Type": "application/json" },
+            statusCode: 400,
+            body: JSON.stringify({
+              error: "invalid_token",
+              error_description: "Missing aud claim",
+            }),
+          });
+        });
+      });
+    });
+  });
+
+  describe("JWT signature verification", () => {
+    describe("Given that the JWT signature verification fails", () => {
+      it("Returns 401 Unauthorized", async () => {
+        dependencies.tokenService = () => new MockTokenSeviceInvalidSignature();
+
+        const event = buildRequest({
+          headers: { Authorization: `Bearer ${mockValidJwt}` },
+        });
+
+        const result: APIGatewayProxyResult = await lambdaHandler(
+          event,
+          dependencies,
+        );
+
+        expect(result).toStrictEqual({
+          headers: { "Content-Type": "application/json" },
+          statusCode: 401,
+          body: JSON.stringify({
+            error: "Unauthorized",
+            error_description: "Invalid signature",
+          }),
+        });
+      });
+    });
+  });
+
+  describe("SSM Service", () => {
+    describe("Given there is an error retrieving client credentials from SSM", () => {
+      it("Returns a 500 Server Error response", async () => {
+        dependencies.ssmService = () => new MockFailingSsmService();
+
+        const event = buildRequest({
+          headers: { Authorization: `Bearer ${mockValidJwt}` },
+        });
+
+        const result = await lambdaHandler(event, dependencies);
+
+        expect(result).toStrictEqual({
+          headers: { "Content-Type": "application/json" },
+          statusCode: 500,
+          body: JSON.stringify({
+            error: "server_error",
+            error_description: "Server Error",
+          }),
+        });
+      });
+    });
+  });
+
+  describe("Client Credentials Service", () => {
```

Comment:
to-do - remove one level of describe block

Reply:
Updated.